Add GranuleListAppendAdapter

lindholc opened the pull request:
Here's what I have so far. Is this generally the kind of thing we want?

I'm ignoring ordering and operations at the moment because the semantics, if they are defined anywhere, are unclear to me.

I'm noticing more reasons that the granule append adapter is a different sort of adapter:
- The `Adapter` trait defines a `getData` method that requires a `URI`. Here we don't have a URI, because the source is the granule list dataset.
- The `AdapterFactory` apply method requires a `DataType`, which this adapter could get from the template. (It might be useful to have a model to check against each application of the template rather than assume the first is correct.)
- It seems likely that this adapter would be constructed specially in `FdmlReader` rather than using `AdapterFactory` because it has special syntax.

lindholc committed:
Add GranuleListAppendAdapter

lindholc commented on core/src/main/scala/latis/input/GranuleListAppendAdapter.scala:
```diff
@@ -0,0 +1,58 @@
+import latis.util.NetUtils
+
+/**
+ * An adapter for creating a dataset from a list of granules.
```
We've established that adapters don't create datasets, they create `Data`. I'm not sure what to say here that wouldn't be confusing. Linking to the `Data` trait doesn't seem good enough.

dlindhol replied:
I started with returning `SampledFunction` since we did not yet have `Data`. I've been planning to change these to `Data` everywhere since there is technically no reason why a dataset can't be a single value of a tuple.

lindholc commented on core/src/main/scala/latis/input/GranuleListAppendAdapter.scala:
```diff
@@ -0,0 +1,58 @@
+ */
+final class GranuleListAppendAdapter(
+  granules: Dataset,
+  template: URI => Dataset
```
I suppose if I made this `URI => SampledFunction`, having the `DataType` from `AdapterFactory` would make more sense.

Is this a reader as it's written?

dlindhol replied:
Hmm. I see a `reader` (not currently well defined) as taking only a `URI` and returning a `Dataset` so the template is effectively a reader. Interesting.
In the sense that this `GranuleListAppendAdapter` provides its own model (from the template), it smells like a `reader`. It needs to be *applied* to a granule list dataset which smells like an `Operation`. This is like what I was trying to do with `ReaderOperation`. However, that brought the need for the `DatasetReader` to provide a model (for the pure `applyToModel` that avoids reading data).

Regarding whether the template should return `Data` instead of `Dataset`:
I think the realization of the template could be treated as a reader. It's unique in that it uses fdml to define it.

I'd like to white-board all these as function signatures and revisit how they might all work together.

lindholc commented on core/src/test/scala/latis/input/GranuleListAppendAdapterSpec.scala:
```diff
@@ -0,0 +1,177 @@
+      glaSamples should equal (gSamples)
+  }
+
+  private val granules: List[List[Sample]] = List(
```
Is there a better way to make these datasets?

dlindhol replied:
I've been building up some utilities in `latis.util.DatasetGenerator` in `src/test`.

lindholc commented on core/src/test/scala/latis/input/GranuleListAppendAdapterSpec.scala:
```diff
@@ -0,0 +1,177 @@
+    new MemoizedDataset(md, model, samples)
+  }
+
+  // A very simple adapter.
```
Similar to a previous comment: is this really reader? Or should it return a `SampledFunction`?

dlindhol commented on core/src/main/scala/latis/input/GranuleListAppendAdapter.scala:
```diff
@@ -0,0 +1,58 @@
+  }
+
+  private def extractUri(s: Sample): Either[LatisException, URI] = s match {
+    case Sample(_, RangeData(Text(uri)))    => NetUtils.parseUri(uri)
```
I'm not sure what we should require about the `uri` variable position. I've considered `projecting` the URI. Or getting the `SamplePosition`.

dlindhol commented on core/src/main/scala/latis/input/GranuleListAppendAdapter.scala:
```diff
@@ -0,0 +1,58 @@
+    getData(ops)
+
+  /** Gets data using this adapter. */
+  def getData(ops: Seq[Operation]): SampledFunction = {
```
private?

lindholc replied:
I didn't make this private because this seems like the method people would actually use rather than the one that forces them to make a fake URI. The other one is there just to satisfy the `Adapter` trait.

dlindhol replied:
I guess I was thinking that the `FdmlReader` would be the only user.

dlindhol reviewed (commented):
I'm still pondering how these pieces might fit together better.

lindholc committed:
Add support for granule appending in FDML

Note this is just adding support for parsing and interpreting FDML
without updating the schema.

lindholc commented:
I've kept the `GranuleListAppendAdapter` largely the same, but it's no longer extending `Adapter`.

I've added support for the new style of FDML, but I don't know what to call the styles. I settled on "dataset FDML" and "granule append FDML" because it was late at night.

Hopefully I'll have time to add a test for the FDML reader. Right now only the parsing and the granule appending itself have tests.

dlindhol commented:
My latest thinking is to use the FDML in its original form and simply allow the `source` to be a granule list dataset (that simply has a `uri` variable that can be projected to result in a sequence of `URIs`). The model would represent a granule (which is equivalent to the final result). The top level dataset definition would include the adapter definition for this `GranuleListAppendAdapter`. (Or just `GranuleAppendAdapter`?)

This will lead to a generalization of `source` to be more than just a `URI`. I image a `DataSource` ADT where one type would be for `URIs` and another for granule list datasets (`GranuleListSource`? `DatasetSource`?). I can also imagine this as a mechanism to inline data values (`InlineDataSource`?). For example, we could include a list of URIs in the fdml file itself! (Or a list of datasets to merge?) This could also be used for things like the TIMED SEE level 4 SSI that does not include wavelength values in the data. 

It feels like a good abstraction, so far. I suspect it means that the `Adapter` API needs to be `DataSource => Data`. I'm not sure if we would need to do  that refactoring here.

lindholc commented:
I'm going to close this for some more work and open another PR when I'm done.
